text/2020-01-20-pre-transfer-leader.md

```diff
@@ -0,0 +1,58 @@
+# Pre-transfer leader
+
+## Summary
+
+[Raft][1] supports leadership transferring by leadership transfer extension.
+This proposal adds an extra step to make leader transferring more smooth.
+
+## Motivation
+
+Leadership transferring may not always succeed. A follower can fail to start
+election because of unapplied conf change or message loss. Even it starts
+election and becomes leader at the end, it may not start to respond to requests
+because logs from last term is still being committed and applied. In the
+practice of TiKV, we observe a slow new leader can take seconds or minutes to
+be active, which is a serious problem needs to be solved.
+
+## Detailed design
+
+To minimize the probability of failure, a leader can send a pre-request to ask
+target follower whether it's suitable to become a leader. The follower can
+check all the situations that can block it from becoming a good leader. If all
+checks pass, follower replies to leader an ack to start actual leadership
+transferring.
+
+To avoid stale ack, pre-request should be attached with leader's current commit
```

Comment:
Do we need a mechanism for pre transfer timeout?

Reply:
I don't think so. The client, pd for example, should retry by itself.